### PR TITLE
Add table aws_route53resolver_rule. Closes #142

### LINKED ISSUE #142: Add table aws_route53resolver_rule
**References**
https://pkg.go.dev/github.com/aws/aws-sdk-go@v1.37.24/service/route53resolver

### PR DESCRIPTION
**Pull request by visiit**
# Integration test logs
<details>
  <summary>Logs</summary>
  
```
vishalkumar@Vishals-MacBook-Air aws-test % TURBOT_PROFILE=shaktiman TF_VAR_aws_profile=shaktiman ./tint.js tests/aws_route53_resolver_rule
No env file present for the current environment:  staging
 Falling back to .env config
No env file present for the current environment:  staging
customEnv TURBOT_TEST_EXPECTED_TIMEOUT undefined

SETUP: tests/aws_route53_resolver_rule []

PRETEST: tests/aws_route53_resolver_rule

TEST: tests/aws_route53_resolver_rule
Running terraform
data.aws_region.alternate: Refreshing state...
data.aws_partition.current: Refreshing state...
data.aws_caller_identity.current: Refreshing state...
data.aws_region.primary: Refreshing state...
data.null_data_source.resource: Refreshing state...
aws_route53_resolver_rule.named_test_resource: Creating...
aws_route53_resolver_rule.named_test_resource: Still creating... [10s elapsed]
aws_route53_resolver_rule.named_test_resource: Creation complete after 16s [id=rslvr-rr-dfe31dc19afa445ab]

Warning: Deprecated Resource

The null_data_source was historically used to construct intermediate values to
re-use elsewhere in configuration, the same can now be achieved using locals


Apply complete! Resources: 1 added, 0 changed, 0 destroyed.

Outputs:

account_id = 986325076436
aws_region = us-east-1
resource_aka = arn:aws:route53resolver:us-east-1:986325076436:resolver-rule/rslvr-rr-dfe31dc19afa445ab
resource_id = rslvr-rr-dfe31dc19afa445ab
resource_name = turbottest68264

Running SQL query: test-get-query.sql
[
  {
    "akas": [
      "arn:aws:route53resolver:us-east-1:986325076436:resolver-rule/rslvr-rr-dfe31dc19afa445ab"
    ],
    "name": "turbottest68264",
    "tags": {
      "name": "turbottest68264"
    },
    "title": "turbottest68264"
  }
]
✔ PASSED

Running SQL query: test-list-query.sql
[
  {
    "id": "rslvr-rr-dfe31dc19afa445ab",
    "name": "turbottest68264",
    "share_status": "NOT_SHARED",
    "status": "COMPLETE",
    "tags": {
      "name": "turbottest68264"
    }
  }
]
✔ PASSED

Running SQL query: test-not-found-query.sql
null
✔ PASSED

Running SQL query: test-turbot-query.sql
[
  {
    "account_id": "986325076436",
    "akas": [
      "arn:aws:route53resolver:us-east-1:986325076436:resolver-rule/rslvr-rr-dfe31dc19afa445ab"
    ],
    "region": "us-east-1",
    "tags": {
      "name": "turbottest68264"
    },
    "title": "turbottest68264"
  }
]
✔ PASSED

POSTTEST: tests/aws_route53_resolver_rule

TEARDOWN: tests/aws_route53_resolver_rule

SUMMARY:

1/1 passed.
```
</details>

# Example query results
<details>
  <summary>Results</summary>
  
```
- List all rules

select
  name,
  domain_name owner_id,
  resolver_endpoint_id,
  rule_type,
  share_status,
  status
from
  aws_route53_resolver_rule;
+-------------------+------------------+-----------------------------+-----------+--------------+----------+
| name              | owner_id         | resolver_endpoint_id        | rule_type | share_status | status   |
+-------------------+------------------+-----------------------------+-----------+--------------+----------+
| Internet Resolver | .                | <null>                      | RECURSIVE | NOT_SHARED   | COMPLETE |
| test-rule         | www.example.com. | rslvr-out-c1515d73fd7a48aa9 | FORWARD   | NOT_SHARED   | COMPLETE |
| Internet Resolver | .                | <null>                      | RECURSIVE | NOT_SHARED   | COMPLETE |
+-------------------+------------------+-----------------------------+-----------+--------------+----------+

- Get a specific rule

select
	name,
	domain_name owner_id,
	resolver_endpoint_id,
	rule_type,
	share_status,
	status
from
	aws_route53_resolver_rule
where
	id = 'rslvr-rr-389d2ef50c092349a';

+-------------+------------------+-----------------------------+-----------+--------------+----------+
| name        | owner_id         | resolver_endpoint_id        | rule_type | share_status | status   |
+-------------+------------------+-----------------------------+-----------+--------------+----------+
| turbot-rule | www.example.com. | rslvr-out-ebb7db0b7498463eb | FORWARD   | NOT_SHARED   | COMPLETE |
+-------------+------------------+-----------------------------+-----------+--------------+----------+

- List all rules where VPCs are not associated

select
  name,
  id,
  arn,
  resolver_rule_associations
from
  aws_route53_resolver_rule
Where
  resolver_rule_associations = '[]';

+----------------+----------------------------+-----------------------------------------------------------------------------------------+----------------------------+
| name           | id                         | arn                                                                                     | resolver_rule_associations |
+----------------+----------------------------+-----------------------------------------------------------------------------------------+----------------------------+
| raj-novpc-rule | rslvr-rr-d397142654d4c6691 | arn:aws:route53resolver:us-east-1:986325076436:resolver-rule/rslvr-rr-d397142654d4c6691 | []                         |
+----------------+----------------------------+-----------------------------------------------------------------------------------------+----------------------------+

- List the IP addresses enabled for outbound DNS queries

select
	name,
	p ->> 'Ip' as ip,
	p ->> 'Port' as port
from
	aws_route53_resolver_rule,
	jsonb_array_elements(target_ips) as p;

+-------------+------------+------+
| name        | ip         | port |
+-------------+------------+------+
| turbot-rule | 192.0.2.10 | 53   |
| turbot-rule | 192.0.2.12 | 53   |
+-------------+------------+------+

-  List of resolver rule shared with another account

select
	name,
	id,
	share_status,
	rule_type
from
	aws_route53_resolver_rule
where
	share_status = 'SHARED'

+------+----+--------------+-----------+
| name | id | share_status | rule_type |
+------+----+--------------+-----------+
+------+----+--------------+-----------+
```
</details>

### CODE REVIEW

File: aws-test/tests/aws_route53_resolver_rule/variables.tf [hunk 1]
```diff
@@ -0,0 +1,76 @@
+variable "resource_name" {
+  type        = string
+  default     = "turbot-test-20200125-create-update"
+  description = "Name of the resource used throughout the test."
+}
+
+variable "aws_profile" {
+  type        = string
+  default     = "integration-tests"
```
Comment (khushboo9024) [inline aws-test/tests/aws_route53_resolver_rule/variables.tf:9]:
```suggestion
  default     = "default"
```

File: aws/table_aws_route53_resolver_rule.go [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+import (
+	"context"
+
+	"github.com/aws/aws-sdk-go/service/route53resolver"
+	"github.com/turbot/go-kit/types"
+	"github.com/turbot/steampipe-plugin-sdk/grpc/proto"
+	"github.com/turbot/steampipe-plugin-sdk/plugin"
+	"github.com/turbot/steampipe-plugin-sdk/plugin/transform"
+)
+
```
Comment (khushboo9024) [inline aws/table_aws_route53_resolver_rule.go:12]:
```suggestion
//// TABLE DEFINITION

```

File: docs/tables/aws_route53_resolver_rule.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+```sql
+select
+	name,
+	p ->> 'Id' as id,
+	p ->> 'Status' as status,
+	p ->> 'VPCId' as vpc_id
+from
+	aws_route53_resolver_rule,
+	jsonb_array_elements(resolver_rule_associations) as p;
```
Comment (khushboo9024) [inline docs/tables/aws_route53_resolver_rule.md:43]: Fix indentation  for all the queries

File: aws/table_aws_route53_resolver_rule.go [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+	return op, nil
+}
+
+//// TRANSFORM FUNCTIONS
+
+func route53resolverRuleTagListToTurbotTags(ctx context.Context, d *transform.TransformData) (interface{}, error) {
+	plugin.Logger(ctx).Trace("route53resolverTagListToTurbotTags")
+
+	if d.Value == nil {
```
Comment (khushboo9024) [inline aws/table_aws_route53_resolver_rule.go:284]: Why we are doing this check , is this required ? Or we should just check the `tagList`.

File: aws/table_aws_route53_resolver_rule.go [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+	}
+
+	// Get call
+	data, err := svc.GetResolverRule(params)
+	if err != nil {
+		logger.Debug("getAwsRoute53ResolverRule", "ERROR", err)
+		return nil, err
+	}
+
+	return data.ResolverRule, nil
```
Comment (khushboo9024) [inline aws/table_aws_route53_resolver_rule.go:200]: Please add a check here for `data.ResolverRule` not equal to nil.

File: aws/table_aws_route53_resolver_rule.go (hunk 4 above)
Comment (visiit) [reply aws/table_aws_route53_resolver_rule.go:284]: this is required. Because default rule doesn,t  suppord tags.

**Review by khushboo9024 — COMMENTED** [review]
Please resolve the conflict

File: aws/plugin.go [hunk 6]
```diff
@@ -103,6 +103,7 @@ func Plugin(ctx context.Context) *plugin.Plugin {
 			"aws_route53_record":                     tableAwsRoute53Record(ctx),
 			"aws_route53_resolver_endpoint":          tableAwsRoute53ResolverEndpoint(ctx),
 			"aws_route53_zone":                       tableAwsRoute53Zone(ctx),
+			"aws_route53_resolver_rule":              tableAwsRoute53ResolverRule(ctx),
```
Comment (khushboo9024) [inline aws/plugin.go:106]: sort the list

File: aws/table_aws_route53_resolver_rule.go [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+			{
+				Name:        "domain_name",
+				Description: "DNS queries for this domain name are forwarded to the IP addresses that are specified in TargetIps.",
+				Type:        proto.ColumnType_STRING,
+			},
+			{
+				Name:        "id",
+				Description: "The ID that Resolver assigned to the Resolver rule when you created it.",
+				Type:        proto.ColumnType_STRING,
+			},
```
Comment (khushboo9024) [inline aws/table_aws_route53_resolver_rule.go:53]: `id` column should be after `name`

**Review by khushboo9024 — DISMISSED** [review]

File: docs/tables/aws_route53_resolver_rule.md [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+  name,
+  domain_name owner_id,
+  resolver_endpoint_id,
+  rule_type,
+  share_status,
+  status
+from
+  aws_route53_resolver_rule
+where
+  id = 'rslvr-rr-389d2ef50c094970b';
```
Comment (rajmohanty17) [inline docs/tables/aws_route53_resolver_rule.md:35]: rslvr-rr-389d2ef50c094970b > rslvr-rr-12345ef50c0949abc

File: docs/tables/aws_route53_resolver_rule.md [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+  name,
+  p ->> 'Ip' as ip,
+  p ->> 'Port' as port
+from
+  aws_route53_resolver_rule,
+  jsonb_array_elements(target_ips) as p;
+```
+
+
+### List of resolver rule not shared with another account
```
Comment (rajmohanty17) [inline docs/tables/aws_route53_resolver_rule.md:66]: List the shared resolver rules.
```sql
select
  name,
  id,
  share_status,
  rule_type
from
  aws_route53_resolver_rule
where
  share_status = 'SHARED';
```

File: docs/tables/aws_route53_resolver_rule.md [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+  p ->> 'Id' as id,
+  p ->> 'Status' as status,
+  p ->> 'VPCId' as vpc_id
+from
+  aws_route53_resolver_rule,
+  jsonb_array_elements(resolver_rule_associations) as p;
+```
+
+
+### List of IP addresses and ports that an outbound endpoint forwards DNS queries
```
Comment (rajmohanty17) [inline docs/tables/aws_route53_resolver_rule.md:53]: List of IP addresses and ports that an outbound endpoint forwards DNS queries >>  List the IP addresses enabled o for outbound DNS queries

File: docs/tables/aws_route53_resolver_rule.md [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+  share_status,
+  status
+from
+  aws_route53_resolver_rule
+where
+  id = 'rslvr-rr-389d2ef50c094970b';
+```
+
+
+### List of associations that were created between Resolver rules and VPCs
```
Comment (rajmohanty17) [inline docs/tables/aws_route53_resolver_rule.md:39]: ### List of associations that were created between Resolver rules and VPCs > List all associated VPCs for the Resolver rules.

We should have a query to find, if we have any resolver rule that is not associated with VPCs. As most of the use case uses VPCs as routing.
  name,
  p ->> 'Id' as id,
  p ->> 'Status' as status,
  p ->> 'VPCId' as vpc_id
from
  aws_route53_resolver_rule,
  jsonb_array_elements(resolver_rule_associations) as p 
where   p ->> 'VPCId' is null ?

**Review by rajmohanty17 — CHANGES_REQUESTED** [review]
Pls check the comments.

File: docs/tables/aws_route53_resolver_rule.md (hunk 8 above)
Comment (visiit) [reply docs/tables/aws_route53_resolver_rule.md]: Resolved

File: docs/tables/aws_route53_resolver_rule.md (hunk 9 above)
Comment (visiit) [reply docs/tables/aws_route53_resolver_rule.md]: Resolved

File: docs/tables/aws_route53_resolver_rule.md (hunk 10 above)
Comment (visiit) [reply docs/tables/aws_route53_resolver_rule.md]: Resolved

File: docs/tables/aws_route53_resolver_rule.md (hunk 11 above)
Comment (visiit) [reply docs/tables/aws_route53_resolver_rule.md]: Resolved

**Review by cbruno10 — APPROVED** [review]

**Pull request merged**